Add Landing page [MAILPOET-4536]

websupporter opened the pull request:
## Description
Adds the different section to the landing page of the automation

## Code review notes
I agree that we should at some point merge automation and automation-templates into one app.

## Linked PRs
https://github.com/mailpoet/mailpoet-premium/pull/640

## Linked tickets
[MAILPOET-4536]

[MAILPOET-4536]: https://mailpoet.atlassian.net/browse/MAILPOET-4536?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

websupporter committed:
Add path the image assets

[MAILPOET-4536]

websupporter committed:
Add <OptionButton> component

[MAILPOET-4536]

websupporter committed:
Add different sections for welcome screen

[MAILPOET-4536]

websupporter committed:
Adjust automation test to new UI

[MAILPOET-4536]

JanJakes commented on mailpoet/assets/js/src/automation/components/option-button/index.tsx:
```diff
@@ -0,0 +1,81 @@
+    document.addEventListener('mousedown', handleCloseOnClick, {
+      capture: true,
+    });
+    document.addEventListener('keyup', handleCloseOnEsc, { capture: true });
```
Why did you choose not to use `DropdownMenu` and implement this manually instead? If we have no strong reason to do our own solution, I'd prefer to use the dropdown menu as it will correctly handle all of the keyboard and accessibility features under all circumstances, OSs, and browsers, for us.

websupporter replied:
The closest to the Figma I cam using the `DropDownMenu` was the following:
![image](https://user-images.githubusercontent.com/6458412/199412548-da653ac0-d2ee-4230-b79b-ba5408f1eb3f.png)

```
    <div className="mailpoet-option-button" ref={ref}>
        <Button
          variant={variant}
          className="mailpoet-option-button-main"
          onClick={onClick}
        >
          {title}
        </Button>
      <DropdownMenu
        className="is-primary"
        label={'...'}
        icon={chevronDown}
        popoverProps={{className:'test',getAnchorRect:() => {
            const element:Element = ref.current;
            return element.getBoundingClientRect();
          }}}
        controls={[{title:'test', icon:chevronUp, isDisabled:false, onClick:() => alert('test')}]} />
    </div>
```

I could not find a good way to determine the width of the menu items to get them to the width of the parent (mostly, because its not the DOM parent)

`popoverProps` allows to send some properties to the popover, like the anchor client rect, but I didn't see something, which could account for width or min-width

JanJakes replied:
@websupporter Hmm, OK, I guess. I know the WP components dropdown isn't great due to the `controls` parameter, but on the other hand, it handles keyboard navigation (also arrows, etc.), it auto-closes on click (we should add that too, so people don't double-click and create multiple workflows), etc.

This seems to work quite OK:
```ts
<style>{`
  .mailpoet-option-button-opener {
    background: var(--wp-admin-theme-color);
    color: white;
  }
  .mailpoet-option-button-opener svg {
    fill: white;
  }
  .mailpoet-option-button-opener .is-opened svg {
    transform: scale(-1, -1);
    transform-origin: center 12.5px;
  }
`}</style>
<DropdownMenu
  className="mailpoet-option-button-opener"
  label={__('More', 'mailpoet')}
  icon={chevronDown}
  controls={[{ title: 'New workflow', icon: undefined }]}
  popoverProps={{ position: 'bottom left' }}
/>
```

Result:

<img width="235" alt="Screenshot 2022-11-02 at 14 28 48" src="https://user-images.githubusercontent.com/141436/199502392-49a09fde-b5a6-41fd-a483-f7a2e1afe877.png">

And the style of the menu below can be CSSed a bit as well. As for the width, I'd omit that, maybe. It's not practical for translations anyway — the button will have different widths in different languages, the dropdown menu items too. Anyway, it's one option, I guess.

Either this or at least adding auto-close on item click.

websupporter replied:
If width is not a problem, DropDownMenu is much better: [5fed668](https://github.com/mailpoet/mailpoet/pull/4461/commits/5fed6688d92b3e05895846c78e23b37df63f349e)

JanJakes commented on mailpoet/assets/css/src/components-automation-listing/_sections.scss:
```diff
@@ -0,0 +1,171 @@
+  margin-left: calc(-50vw + var(--menu-width) / 2 + 50%);
+  padding-left: 104px;
+  padding-right: 104px;
+  width: calc(100vw - var(--menu-width));
```
You seem to be setting `100vh` and then subtracting the (expected) menu, is that correct? Is there a reason for that? It seems to be unnecessarily complex to me, and replicating things like `min-width: 783px` and `--menu-width: 36px` is not nice either. I think this may also break if some plugins change the admin styles, etc.

So now the "why don't you just" question 😄 Why don't you just fill the whole content part with sth like `width: 100%`?

websupporter replied:
bad67451a

websupporter replied:
After rebase: [19b5349](https://github.com/mailpoet/mailpoet/pull/4461/commits/19b5349a9808bebf5124fa7774152fb8c4c4cabf)

JanJakes commented on mailpoet/assets/js/src/automation/sections/templates-section.tsx:
```diff
@@ -0,0 +1,33 @@
+      <h2>{__('Start with essentials', 'mailpoet')}</h2>
+      <p>
+        {__(
+          'Choose from our list of pre-made templates and make it your own',
```
A trailing `.` is missing.

websupporter replied:
c4f2431fd

websupporter replied:
After rebase: [c437320](https://github.com/mailpoet/mailpoet/pull/4461/commits/c43732067d3e0af0e1dccb831c0c226b0314d2b0)

JanJakes commented on mailpoet/assets/js/src/automation/sections/templates-section.tsx:
```diff
@@ -0,0 +1,33 @@
+
+  return (
+    <section className="mailpoet-automation-section mailpoet-section-templates">
+      <h2>{__('Start with essentials', 'mailpoet')}</h2>
```
In Figma, I see "Explore essentials", or was that changed somewhere?

websupporter replied:
current Figma is [this](https://www.figma.com/file/SAeK2GqSPbGsO7j5iQPT1j/Automations?node-id=1919%3A298068), right?
![image](https://user-images.githubusercontent.com/6458412/199194677-a6fa9581-196b-4e93-8653-68a2ec9e8a9f.png)

I think the screenshot in the issue is outdated.

JanJakes replied:
@websupporter Ahhh, there are both strings in Figma, depending on which of those screens you look. @NeosinneR which one is correct?

NeosinneR replied:
[This](https://www.figma.com/file/SAeK2GqSPbGsO7j5iQPT1j/Automations?node-id=1919%3A298068) should be the latest version - so "Explore essentials".

websupporter replied:
684aad2fe

websupporter replied:
After rebase: [c92e827](https://github.com/mailpoet/mailpoet/pull/4461/commits/c92e82786150d9e77ad1a49a28bfad613100505f)

JanJakes commented on mailpoet/assets/js/src/automation/sections/templates-section.tsx:
```diff
@@ -0,0 +1,33 @@
+      </p>
+      <ul className="mailpoet-section-template-list">
+        {templates.map((template) => (
+          <TemplateListItem
```
For some reason those don't seem to have the same hover effect as templates do.

websupporter replied:
dfee86273

websupporter replied:
After rebase: [0cbddaf](https://github.com/mailpoet/mailpoet/pull/4461/commits/0cbddafb899ce53abe1f4ba515ce920b6167352e)

JanJakes commented on mailpoet/assets/js/src/automation/templates/components/template-list-item.tsx:
```diff
@@ -50,7 +53,9 @@ export function TemplateListItem({
       >
-        <h2>{template.name} →</h2>
+        {headingTag === 'h3' && <h3>{template.name} →</h3>}
+        {headingTag === 'h2' && <h2>{template.name} →</h2>}
```
Should we put a `&nbsp;` before the `→`?

<img width="394" alt="Screenshot 2022-10-27 at 9 31 57" src="https://user-images.githubusercontent.com/141436/198219450-2305fb89-9cee-413f-a3da-1dc3fa57f29e.png">

websupporter replied:
3dae2d4b0

websupporter replied:
After rebase: [95e1290](https://github.com/mailpoet/mailpoet/pull/4461/commits/95e12908bd2a28c2b6ee987567812c544f93d401)

JanJakes commented on mailpoet/lib/AdminPages/Pages/Automation.php:
```diff
@@ -43,6 +50,12 @@ public function render() {
       'workflowCount' => $this->workflowStorage->getWorkflowCount(),
+      'templates' => array_map(
+        function(WorkflowTemplate $template): array {
+          return $template->toArray();
```
Ideally, we should create a template mapper as `toArray` is mainly for the DB. But good for now.

JanJakes commented on mailpoet/assets/css/src/components-automation-listing/_sections.scss:
```diff
@@ -0,0 +1,171 @@
+  .mailpoet-section-template-list {
+    display: flex;
+    justify-content: space-between;
+    margin-bottom: 40px;
```
The boxes can have different heights now, I think it would be nicer to make them same-height (e.g. by CSS grid).

<img width="1472" alt="Screenshot 2022-10-27 at 9 35 00" src="https://user-images.githubusercontent.com/141436/198220007-144299fb-5447-4624-8ba9-b224e4961c7b.png">

NeosinneR replied:
👍 and also always have some space between the boxes. Now on smaller displays, they are squashed together.

websupporter replied:
Ensure space between: 8b8811e58
Same height: 50ff58b45

websupporter replied:
After rebase:
space between: [c543b1a](https://github.com/mailpoet/mailpoet/pull/4461/commits/c543b1a313992a1c10fc42fe26b67bd01cb5e784)
same height: [496d616](https://github.com/mailpoet/mailpoet/pull/4461/commits/496d6162668e312fb89b6240fd758067addc3b25)

JanJakes replied:
Looks good. Could we also ensure the same width and align the content to the top? What I mean is:

<img width="804" alt="Screenshot 2022-11-02 at 13 39 29" src="https://user-images.githubusercontent.com/141436/199491764-ad0afd4c-f066-45cc-a06a-ccd5670f4386.png">

websupporter replied:
Good catch. Thank you!

4bb6330a5 and 9b6a0db01 ensure now same width and top alignment.

JanJakes commented on mailpoet/assets/css/src/components-automation-listing/_sections.scss:
```diff
@@ -0,0 +1,171 @@
+    max-width: 373px;
+    padding: 0;
+
+    .mailpoet-section-build-list-button {
```
I see you're using a lot of nesting. It can become a bit hard to read, I'd say. We don't have any strict rules here, but I think we tend to put all unique-enough "FQN" qualifiers (like custom classes, e.g. `.mailpoet-section-build-list-button`) to the top level.

websupporter replied:
Addressed in [7345003](https://github.com/mailpoet/mailpoet/pull/4461/commits/7345003b7bccf30ca7e17b1cfb9643b6ddac7450).

I actually like it "a bit nested". It groups the components CSS together, but can be too much, I agree. Will try to follow the teams standards.

websupporter replied:
After rebase [d593be6](https://github.com/mailpoet/mailpoet/pull/4461/commits/d593be642f7b483a012a9bbb472b0941cca1181c)

JanJakes commented on mailpoet/assets/js/src/automation/sections/build-your-own-section.tsx:
```diff
@@ -0,0 +1,79 @@
+import { __ } from '@wordpress/i18n';
+import { MailPoet } from '../../mailpoet';
+
+export function BuildYourOwnSection(): JSX.Element {
```
This seems to be somewhat misaligned in my browser (Chrome, macOS) — the underline as well as heading vs. details text:

<img width="539" alt="Screenshot 2022-10-27 at 9 45 33" src="https://user-images.githubusercontent.com/141436/198222264-26f0ef60-d583-4830-8f9d-1f7b4982748b.png">



Specs:
<img width="734" alt="Screenshot 2022-10-27 at 9 44 36" src="https://user-images.githubusercontent.com/141436/198222064-c41faf0b-e2e1-4bbb-9eae-60c8bd5978e1.png">

JanJakes replied:
Also the spacing below the item here seems too large:

<img width="473" alt="Screenshot 2022-10-27 at 9 51 02" src="https://user-images.githubusercontent.com/141436/198223386-8a9a304b-cf5e-4819-a480-e80763b1a12a.png">

websupporter replied:
58e9532e4

websupporter replied:
After rebase: [be69f35](https://github.com/mailpoet/mailpoet/pull/4461/commits/be69f355fd253c86da311b916b8efd26566d950d)

JanJakes commented on mailpoet/lib/AdminPages/PageRenderer.php:
```diff
@@ -155,6 +156,7 @@ public function displayPage($template, array $data = []) {
         'automationListing' => admin_url('admin.php?page=mailpoet-automation'),
         'automationEditor' => admin_url('admin.php?page=mailpoet-automation-editor'),
         'automationTemplates' => admin_url('admin.php?page=mailpoet-automation-templates'),
+        'imageAssets' => plugin_dir_url(Env::$file) . 'assets/img/',
```
I think we store any bigger assets under `plugin_repository/assets`, which pushes them to WP.org CDN.

That said, do we have these as SVGs? I know there's a photo in one of them, so it can be problematic. In all cases, they look blurry on a retina display so even with PNGs, we'll need a retina-friendly resolution, I guess.

As for the path, in code we seem to namespace it `automation`, not `automations`.

JanJakes replied:
I just check Figma, and they seem to be SVGs, so let's use that, if possible 🙏

websupporter replied:
I downloaded the SVGs and because of the embedded images and base64, they grow up to 10MB. I used 2x PNGs instead: 9036139d7

Assets path change: f0bcb92c1

websupporter replied:
after rebase
images: [f136511](https://github.com/mailpoet/mailpoet/pull/4461/commits/f136511fd82930118c214217802524ce6e2abf13)
asset path: [9e1df17](https://github.com/mailpoet/mailpoet/pull/4461/commits/9e1df17328c4c4749e7fe07aaf9c81666c3c51f4)

JanJakes replied:
> I think we store any bigger assets under `plugin_repository/assets`, which pushes them to WP.org CDN.

I guess this part got lost. In case we have to keep PNGs, I guess we may want to do that.

websupporter replied:
[b425650](https://github.com/mailpoet/mailpoet/pull/4461/commits/b4256500773147b102c1a17854f8e0758150d0eb) uses CDN path now.

JanJakes commented:
One more observation — I see the landing page even on the listing page (below it).

JanJakes commented:
Oh, and also adding here: https://a8c.slack.com/archives/C0313447YBC/p1666854722129479

websupporter committed:
Add missing period

[MAILPOET-4536]

websupporter committed:
align template-list items layout

[MAILPOET-4536]

websupporter committed:
Rephrase to 'Explore essentials'

[MAILPOET-4536]

websupporter committed:
Add Non-breaking space

[MAILPOET-4536]

websupporter committed:
Ensure template items are of same height

[MAILPOET-4536]

websupporter committed:
Simplify full-width definition

The values to define fullwidth are derived from #wpcontent and .wrap
The div.wrap exists actually twice in our templates, so those values
needed to be multiplied by 2.

[MAILPOET-4536]

websupporter committed:
Use images with a better quality

[MAILPOET-4536]

websupporter committed:
Move automation image assets directory

[MAILPOET-4536]

websupporter committed:
Reduce right padding of sections by 20px

Those 20px to the border are already in the double usage of the .wrap class in the DOM

[MAILPOET-4536]

websupporter committed:
Ensure there is space between template items and enable wrap

[MAILPOET-4536]

websupporter committed:
Add space between options menu and options button

[MAILPOET-4536]

websupporter committed:
Adjust layout of build-your-own list

[MAILPOET-4536]

websupporter committed:
Let image go below build-your-own list on mobile

[MAILPOET-4536]

websupporter committed:
Hide notices on onboarding screen

[MAILPOET-4536]

websupporter committed:
Reduce nesting in CSS

[MAILPOET-4536]

websupporter commented:
Thank you @JanJakes for your review. I addressed most issues

> One more observation — I see the landing page even on the listing page (below it).

Doesn't it follow from the requirements 2.a.i, 2.b.i and 2.c.i in the ticket?
Mainly those say for the template and build your own section: "This section is always visible, also below the workflow listing."

> Oh, and also adding here: https://a8c.slack.com/archives/C0313447YBC/p1666854722129479

[606c23f](https://github.com/mailpoet/mailpoet/pull/4461/commits/606c23fd4d79cfbc1018836fdb4c30590d352946)

websupporter committed:
Add automation listing header

[MAILPOET-4536]

websupporter committed:
Sort use statements alphabetically

[MAILPOET-4536]

JanJakes commented on mailpoet/assets/css/src/components-automation-listing/_listing.scss:
```diff
@@ -6,6 +6,12 @@
 }
 
+.mailpoet-automation-is-onboarding {
+  .notice {
```
I think we'll need to include more classes, not sure how many exist but I can see an `error`:

<img width="557" alt="Screenshot 2022-11-02 at 13 32 42" src="https://user-images.githubusercontent.com/141436/199490554-f4e29578-3e00-4be9-9c46-27e903d1a3fd.png">

JanJakes replied:
Ahh, I see this is fixed here: https://github.com/mailpoet/mailpoet-premium/pull/646/files

websupporter committed:
Use cdn url for automation assets

[MAILPOET-4536]

websupporter committed:
Add mailpoet-automation-section-content to define a max-width for section content

[MAILPOET-4536]

websupporter committed:
Use DropdownMenu in OptionButton component

[MAILPOET-4536]

websupporter committed:
Keep template button columns aligned on top

[MAILPOET-4536]

websupporter committed:
Ensure same width of template buttons

[MAILPOET-4536]

websupporter committed:
Adjust hight of hero picture to align with blue button

[MAILPOET-4536]

websupporter committed:
Improve mobile behavior of hero image

[MAILPOET-4536]
